### ISSUE TITLE
Feature Proposal: Orderability of Routes 

### ISSUE THREAD

**Issue by superdummy** [laravel/framework#10891]
Hi, 
laravel is great and the package system also allows to develop good modular applications, super! 

But currently the package system has one big deficiency that routes (added by a Package's ServiceProvider) can not be "positioned" in the list of routes. What would be required for a lot of packages (like cms'es) and use cases. For example having a "Catch All Route" is a must in lots of applications and the current system does not allow to have a working "Catch All Route" defined (either by the application itself or by a package).. And in general, orderability or routes is a must for any real plattform/framework.

Someone having the problem too (I am not the only one): https://laracasts.com/discuss/channels/laravel/cms-catch-all-route-conflicting-with-serviceprovider-routes

Proposal:
Adding a further optional parmater like "order" into the route methods parameter array like:
`Route::get('foo/{slug}', ['order'=>100, 'uses=>''PostController@show');`
To avoid any performance hit for sorting, when the parameter array is parsed and a order parameter is found, then set an internal flag, that ordering is enabled/required. So only if one uses the features (by setting at least one 'order'=>) theres the performance penality of sorting.

(Or as fallback, at least allow to specify some positions like "top" or "bottom" (on the route list). It at least allows for one package to deterministically put a rule at top or bottom, what will most likely be the most important use-case. However ordering would be much more versatile und usefull). 

I think this would add great benefit to laravel as a plattform (even wordpress has this, although WP does not have a public router like laravel)...

Thanks!

**Comment by rkgrep** [comment]
This will cause a complete mess in developer's routes files (consider working in a non-perfect team or with some novice who didn't understand original routes logic). And at some time in a big project (200+ routes) it will cause a great headache to find some route that is blocking other.

The reason that WP has this is that there may be a lot of modules installed which route ordering cannot be controlled (because people who use it sometimes even don't know how it works).

**Comment by superdummy** [comment]
thanks for your comment. Yes this is in important aspect but can be very simply resovled by effectively only sorting those, that have a number. Everything else is kept in the order it was added (this is indeed an important requirement.)

I was really like the packages architecture and started to "Modularize" my whole application. I think this is a very good practice for Laravel too. And would very much improve the eco-system with more modules (I built a cms and every feature will now be a module.  I plan to release them also as standalone OSS package. So I think more development in laravel should be made in packages not monolithic big apps). And in this regard WP and laravel will then be the same, having lots of packages and this logic is quite important then.
